### ISSUE TITLE
Bug: SerializationException with [...]*Struct (fixed size array of pointers containing nils)

### ISSUE THREAD

**Issue by finnbear** [guregu/dynamo#137]
I am getting
```
2020/08/03 17:55:30 Save error: SerializationException: 
	status code: 400, request id: 4I7AF5B849G17U0MTIIR89EHTJVV4KQNSO5AEMVJF66Q9ASUAAJG
```
when I try to issue a PUT request regarding this struct:
```go
    Floor struct {
        Fences   [6]*Fence `dynamo:"fences"`
    }

    // Old version of struct that doesn't suffer from the problem but has exact same memory layout
    FloorThatWorkedBefore struct {
        Fence0   *Fence    `json:"fence0,omitempty"`
        Fence1   *Fence    `json:"fence1,omitempty"`
        Fence2   *Fence    `json:"fence2,omitempty"`
        Fence3   *Fence    `json:"fence3,omitempty"`
        Fence4   *Fence    `json:"fence4,omitempty"`
        Fence5   *Fence    `json:"fence5,omitempty"`
    }
```
By modifying this library's code to print the request as JSON, I identified the issue as the following:
```json
            "fences": {
                  "B": null,
                  "BOOL": null,
                  "BS": null,
                  "L": [
                    null, // HERE
                    {
                      "B": null,
                      "BOOL": null,
                      "BS": null,
                      "L": null,
                      "M": {
                        "created": {
                          "B": null,
                          "BOOL": null,
                          "BS": null,
                          "L": null,
                          "M": null,
                          "N": "1596477319786",
                          "NS": null,
                          "NULL": null,
                          "S": null,
                          "SS": null
                        }
                      },
                      "N": null,
                      "NS": null,
                      "NULL": null,
                      "S": null,
                      "SS": null
                    },
                    null, // HERE
                    null, // HERE
                    null, // HERE
                    {
                      "B": null,
                      "BOOL": null,
                      "BS": null,
                      "L": null,
                      "M": {
                        "created": {
                          "B": null,
                          "BOOL": null,
                          "BS": null,
                          "L": null,
                          "M": null,
                          "N": "1596477321469",
                          "NS": null,
                          "NULL": null,
                          "S": null,
                          "SS": null
                        },
                      },
                      "N": null,
                      "NS": null,
                      "NULL": null,
                      "S": null,
                      "SS": null
                    }
                  ],
                  "M": null,
                  "N": null,
                  "NS": null,
                  "NULL": null,
                  "S": null,
                  "SS": null
                },
```
If it isn't obvious, the nulls in the list were not encoded correctly. In dynamo JSON, all nulls must be as follows:
```json
          {
            "NULL": true
          },
```
Making the correct version of the list (excuse the incorrect indentation):
```json
    "floors": {
        "L": [
          {
            "NULL": true
          },
          {
                      "B": null,
                      "BOOL": null,
                      "BS": null,
                      "L": null,
                      "M": {
                        "created": {
                          "B": null,
                          "BOOL": null,
                          "BS": null,
                          "L": null,
                          "M": null,
                          "N": "1596477319786",
                          "NS": null,
                          "NULL": null,
                          "S": null,
                          "SS": null
                        }
                      },
                      "N": null,
                      "NS": null,
                      "NULL": null,
                      "S": null,
                      "SS": null
                    },
          {
            "NULL": true
          },
          {
            "NULL": true
          },
          {
            "NULL": true
          },
          {
                      "B": null,
                      "BOOL": null,
                      "BS": null,
                      "L": null,
                      "M": {
                        "created": {
                          "B": null,
                          "BOOL": null,
                          "BS": null,
                          "L": null,
                          "M": null,
                          "N": "1596477321469",
                          "NS": null,
                          "NULL": null,
                          "S": null,
                          "SS": null
                        },
                      },
                      "N": null,
                      "NS": null,
                      "NULL": null,
                      "S": null,
                      "SS": null
                    }
        ]
      }
```
I'm guessing the problem is that, while the decoder handles NULLs:
https://github.com/guregu/dynamo/blob/030e5eaaa355cc89d69099f43ba1fd9fac5fa3aa/decode.go#L81-L84
The encoder only uses nils:
https://github.com/guregu/dynamo/blob/030e5eaaa355cc89d69099f43ba1fd9fac5fa3aa/encode.go#L292-L299
https://github.com/guregu/dynamo/blob/030e5eaaa355cc89d69099f43ba1fd9fac5fa3aa/encode.go#L190-L196

**Comment by guregu** [comment]
Hello, thanks for the detailed report. I'll try and get this and #102 fixed soon, they are related.
In the meantime you could define a type for `[6]*Fence` and write a custom marshaling method for it, but that's just a stopgap solution.
First I'll take care of your PR, sorry for the delays. I've been busy lately.

**Comment by finnbear** [comment]
Thank you! Note that I'd much rather have this fixed as it affects the format of my database which I am hoping the solidify asap. If you think it would be faster to give me a hint about how to fix it, I could try formulating a PR. #133 only makes my Go code *slightly* nicer; it's less of a priority to me.

**Comment by guregu** [comment]
This is fixed in `v1.9.0`, just released.
Nils inside of slices and arrays will be encoded to a DynamoDB NULL type now. 
Thanks for the issue. Let me know if you have any problems with the fix.
